src/inmanta/server/services/paramservice.py

```diff
@@ -200,54 +221,85 @@ async def _update_param(
+    def _validate_parameter(
+        self,
+        name: str,
+        resource_id: Optional[str],
+        expires: Optional[bool],
+    ) -> None:
+        if not resource_id and expires:
+            # Parameters cannot expire
+            raise BadRequest(
+                "Cannot update or set parameter %s: `expire` set to True but parameters cannot expire."
+                "Consider using a fact instead by providing a resource_id." % name,
+            )
+
+    @handle(methods.set_param, name="id", env="tid")
     async def set_param(
         self,
         env: data.Environment,
-        param_id: str,
+        name: str,
         source: ParameterSource,
         value: str,
-        resource_id: str,
+        resource_id: Optional[str],
         metadata: JsonType,
         recompile: bool,
+        expires: Optional[bool] = None,
     ) -> Apireturn:
-        result = await self._update_param(env, param_id, value, source, resource_id, metadata, recompile)
+        self._validate_parameter(name, resource_id, expires)
+        result = await self._update_param(env, name, value, source, resource_id, metadata, recompile, expires)
         warnings = None
         if result:
             compile_metadata = {
                 "message": "Recompile model because one or more parameters were updated",
                 "type": "param",
-                "params": [(param_id, resource_id)],
+                "params": [(name, resource_id)],
             }
             warnings = await self.server_slice._async_recompile(env, False, metadata=compile_metadata)
 
         if resource_id is None:
             resource_id = ""
 
-        params = await data.Parameter.get_list(environment=env.id, name=param_id, resource_id=resource_id)
+        params = await data.Parameter.get_list(environment=env.id, name=name, resource_id=resource_id)
 
         return attach_warnings(200, {"parameter": params[0]}, warnings)
 
     @handle(methods.set_parameters, env="tid")
     async def set_parameters(self, env: data.Environment, parameters: list[dict[str, Any]]) -> Apireturn:
         recompile = False
 
+        updating_facts: bool = False
+        updating_parameters: bool = False
+        parameters_and_or_facts: str = "parameters"
+
         params: list[tuple[str, ResourceIdStr]] = []
+
         for param in parameters:
             name: str = param["id"]
             source = param["source"]
             value = param["value"] if "value" in param else None
             resource_id: ResourceIdStr = param["resource_id"] if "resource_id" in param else None
             metadata = param["metadata"] if "metadata" in param else None
-            expires = param["expires"] if "expires" in param else True
+            expires = param["expires"] if "expires" in param else None
 
-            result = await self._update_param(env, name, value, source, resource_id, metadata, expires)
+            self._validate_parameter(name, resource_id, expires)
```

Comment:
We should move this validation up, so that we refuse the entire request if there is one invalid parameter in the list. Before applying any changes.